server/migrator/Exporter.java

```diff
@@ -0,0 +1,197 @@
+    }
+
+    private DataProto.Item readRelation(Relation relation) {
+        relationCount.incrementAndGet();
+        DataProto.Item.Relation.Builder relationBuilder = DataProto.Item.Relation.newBuilder()
+                .setId(new String(relation.getIID()))
+                .setLabel(relation.getType().getLabel().name());
+        Map<? extends RoleType, ? extends List<? extends Thing>> playersByRole = relation.getPlayersByRoleType();
+        for (Map.Entry<? extends RoleType, ? extends  List<? extends Thing>> rolePlayers : playersByRole.entrySet()) {
+            RoleType role = rolePlayers.getKey();
+            DataProto.Item.Relation.Role.Builder roleBuilder = DataProto.Item.Relation.Role.newBuilder()
+                    .setLabel(role.getLabel().scopedName());
+            for (Thing player : rolePlayers.getValue()) {
+                playerCount.incrementAndGet();
+                roleBuilder.addPlayer(DataProto.Item.Relation.Role.Player.newBuilder()
+                        .setId(new String(player.getIID())));
+            }
+            relationBuilder.addRole(roleBuilder);
+        }
+        readOwnerships(relation).forEach(relationBuilder::addAttribute);
+        return DataProto.Item.newBuilder().setRelation(relationBuilder).build();
+    }
+
+    private DataProto.Item readAttribute(Attribute attribute) {
+        attributeCount.incrementAndGet();
+        DataProto.Item.Attribute.Builder attributeBuilder = DataProto.Item.Attribute.newBuilder()
+                .setId(new String(attribute.getIID()))
+                .setLabel(attribute.getType().getLabel().name())
+                .setValue(readValue(attribute));
+        readOwnerships(attribute).forEach(attributeBuilder::addAttribute);
+        return DataProto.Item.newBuilder().setAttribute(attributeBuilder).build();
+    }
+
+    private DataProto.ValueObject.Builder readValue(Attribute attribute) {
+        DataProto.ValueObject.Builder valueObject = DataProto.ValueObject.newBuilder();
+        if (attribute instanceof Attribute.String) {
+            valueObject.setString(attribute.asString().getValue());
+        } else if (attribute instanceof Attribute.Boolean) {
+            valueObject.setBoolean(attribute.asBoolean().getValue());
+        } else if (attribute instanceof Attribute.Long) {
+            valueObject.setLong(attribute.asLong().getValue());
+        } else if (attribute instanceof Attribute.Double) {
+            valueObject.setDouble(attribute.asDouble().getValue());
+        } else if (attribute instanceof Attribute.DateTime) {
+            valueObject.setDatetime(attribute.asDateTime().getValue().atZone(ZoneId.of("Z")).toInstant().toEpochMilli());
+        } else {
+            throw GraknException.of(ILLEGAL_STATE);
+        }
+        return valueObject;
+    }
+
+    private Stream<DataProto.Item.OwnedAttribute.Builder> readOwnerships(Thing thing) {
+        return thing.getHas().map(attribute -> {
+            ownershipCount.incrementAndGet();
```

Comment:
I'm not a huge fan of doing this `incrementAndGet` inside a `map` lambda expression. It could have unexpected side effects if, say, the code were changed to call `readOwnerships` multiple times. As a rule of thumb, performing state changes inside a `map` is dangerous.

Could we move it outside the map - perhaps into `readEntity`, `readAttribute` and `readRelation`? I know it would mean some code duplication but I think the flow would be clearer and safer.

Reply:
Nah, this needs to be incremented for each edge, will see what I can do there.

Reply:
Ok moving them into `readEntity/...` seems to be the easiest.